core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -525,21 +524,22 @@ private void applyNewOrUpdatedShards(final ClusterChangedEvent event) {
                 assert currentRoutingEntry.isSameAllocation(shardRouting) :
                     "local shard has a different allocation id but wasn't cleaning by applyDeletedShards. "
                         + "cluster state: " + shardRouting + " local: " + currentRoutingEntry;
-                if (isPeerRecovery(shardRouting)) {
-                    final DiscoveryNode sourceNode = findSourceNodeForPeerRecovery(routingTable, nodes, shardRouting);
-                    // check if there is an existing recovery going, and if so, and the source node is not the same, cancel the recovery to restart it
-                    if (recoveryTargetService.cancelRecoveriesForShard(indexShard.shardId(), "recovery source node changed", status -> !status.sourceNode().equals(sourceNode))) {
-                        logger.debug("[{}][{}] removing shard (recovery source changed), current [{}], global [{}])", shardRouting.index(), shardRouting.id(), currentRoutingEntry, shardRouting);
-                        // closing the shard will also cancel any ongoing recovery.
-                        indexService.removeShard(shardRouting.id(), "removing shard (recovery source node changed)");
-                        shardHasBeenRemoved = true;
+                if (shardRouting.isPeerRecovery()) {
+                    RecoveryState recoveryState = indexShard.recoveryState();
+                    if (recoveryState != null && recoveryState.getStage() != RecoveryState.Stage.DONE) { // shard is still recovering
+                        if (recoveryState.getType() == RecoveryState.Type.PRIMARY_RELOCATION || recoveryState.getType() == RecoveryState.Type.REPLICA) {
+                            final DiscoveryNode sourceNode = findSourceNodeForPeerRecovery(logger, routingTable, nodes, shardRouting);
+                            if (recoveryState.getSourceNode().equals(sourceNode) == false) {
```

Comment:
> can be cancelled just because primary relocation completed before shard was activated by the master node

yes. I'm aware of that - I'm thinking that with seq# fast recovery it wouldn't matter much as a ready shard will quickly re-recover. However, seeing how the new code looks like with the cancelRecoveriesForShard + shardRouting.isPeerRecovery changes, I think it became much simpler. I'm good. We can see how things develop later on and potentially move some logic to the master (which will simplify this class) or not.